shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java

```diff
@@ -0,0 +1,247 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugin.editing;
+
+import android.text.Editable;
+import android.text.Selection;
+import android.text.SpannableStringBuilder;
+import android.view.View;
+import android.view.inputmethod.BaseInputConnection;
+import io.flutter.Log;
+import io.flutter.embedding.engine.systemchannels.TextInputChannel;
+import java.util.ArrayList;
+
+/// The current editing state (text, selection range, composing range) the text input plugin holds.
+///
+/// As the name implies, this class also notifies its listeners when the editing state changes. When
+/// there're ongoing batch edits, change notifications will be deferred until all batch edits end
+/// (i.e. when the outermost batch edit ends). Listeners added during a batch edit will always be
+/// notified when all batch edits end, even if there's no real change.
+///
+/// Adding/removing listeners or changing the editing state in a didChangeEditingState callback may
+/// cause unexpected behavior.
+//
+// Currently this class does not notify its listeners on spans-only changes (e.g.,
+// Selection.setSelection). Wrap them in a batch edit to trigger a change notification.
+class ListenableEditingState extends SpannableStringBuilder {
+  interface EditingStateWatcher {
+    // Changing the editing state in a didChangeEditingState callback may cause unexpected
+    // behavior.
+    void didChangeEditingState(
+        boolean textChanged, boolean selectionChanged, boolean composingRegionChanged);
+  }
+
+  private static final String TAG = "flutter";
```

Comment:
We seem to use a tag specific to the class name in some cases, but (don't quote me on this) the tool may filter for `flutter` explicitly, so you should probably verify the message is visible when you want it to be.

Reply:
Updated all 3 classes to use their class names.

Reply:
If I change the level to `Log.v` it doesn't show up even if I change the `TAG` to "flutter" and use `flutter run -v`.